.github/workflows/code_formatting.yml

```diff
@@ -1,16 +1,10 @@
-name: Check code formatting
-
+# https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-python
+name: Python code formatting with black
 on: [push, pull_request]
-
 jobs:
-  build:
+  black:
     runs-on: ubuntu-latest
     steps:
-    - uses: actions/checkout@v3
-    - uses: actions/setup-python@v4
-    - name: Install packages
-      run: source tools/ci.sh && ci_code_formatting_setup
-    - name: Run code formatting
-      run: source tools/ci.sh && ci_code_formatting_run
-    - name: Check code formatting
-      run: git diff --exit-code
+    - uses: actions/checkout@v4
+    - run: pip install --user black==23.7.0
```

Comment:
Maybe I have too much DRY brainworms, but could we add a `[project.optional-dependencies]` section to pyproject.toml for these versions? (I/we could also do this in a follow-up PR for this & ruff together, if that's easier.)

The main advantage is that we could then tell developers to install the linter tools from there (and Python developers will instinctively go there anyway). This means they will have matching versions in any IDE integrations, pylsp, etc and [hopefully] won't run into the situation of the pre-commit hook re-formatting (or re-linting) code that their IDE had already formatted or linted slightly differently.

(AFAIK the version still needs to be duplicated in two places even then, because pre-commit maintains its own sandboxes and reads those versions `.pre-commit-config.yaml`. No way around that, from what I can see.)